### PR TITLE
pythonPackages.ansible-core: fix ansible-galaxy

### PR DESCRIPTION
**Pull request by stephank**
###### Motivation for this change

```
$ ansible-galaxy collection install -r requirements.yml -p collections -vvv
ansible-galaxy [core 2.11.2]
  config file = REDACTED
  configured module search path = ['/Users/stephank/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /nix/store/ipn0x2skyksfmmqv3z4dkzff5jz1gxbm-python3.8-ansible-core-2.11.2/lib/python3.8/site-packages/ansible
  ansible collection location = REDACTED
  executable location = /nix/store/ipn0x2skyksfmmqv3z4dkzff5jz1gxbm-python3.8-ansible-core-2.11.2/bin/ansible-galaxy
  python version = 3.8.9 (default, May 27 2021, 03:09:34) [Clang 7.1.0 (tags/RELEASE_710/final)]
  jinja version = 2.11.3
  libyaml = True
Using REDACTED as config file
Reading requirement file at 'REDACTED/requirements.yml'
Starting galaxy collection install process
Process install dependency map
ERROR! Unexpected Exception, this is probably a bug: find_matches() got an unexpected keyword argument 'identifier'
the full traceback was:

Traceback (most recent call last):
  File "/nix/store/ipn0x2skyksfmmqv3z4dkzff5jz1gxbm-python3.8-ansible-core-2.11.2/bin/.ansible-galaxy-wrapped", line 136, in <module>
    exit_code = cli.run()
  File "/nix/store/ipn0x2skyksfmmqv3z4dkzff5jz1gxbm-python3.8-ansible-core-2.11.2/lib/python3.8/site-packages/ansible/cli/galaxy.py", line 552, in run
    return context.CLIARGS['func']()
  File "/nix/store/ipn0x2skyksfmmqv3z4dkzff5jz1gxbm-python3.8-ansible-core-2.11.2/lib/python3.8/site-packages/ansible/cli/galaxy.py", line 75, in method_wrapper
    return wrapped_method(*args, **kwargs)
  File "/nix/store/ipn0x2skyksfmmqv3z4dkzff5jz1gxbm-python3.8-ansible-core-2.11.2/lib/python3.8/site-packages/ansible/cli/galaxy.py", line 1186, in execute_install
    self._execute_install_collection(
  File "/nix/store/ipn0x2skyksfmmqv3z4dkzff5jz1gxbm-python3.8-ansible-core-2.11.2/lib/python3.8/site-packages/ansible/cli/galaxy.py", line 1213, in _execute_install_collection
    install_collections(
  File "/nix/store/ipn0x2skyksfmmqv3z4dkzff5jz1gxbm-python3.8-ansible-core-2.11.2/lib/python3.8/site-packages/ansible/galaxy/collection/__init__.py", line 511, in install_collections
    dependency_map = _resolve_depenency_map(
  File "/nix/store/ipn0x2skyksfmmqv3z4dkzff5jz1gxbm-python3.8-ansible-core-2.11.2/lib/python3.8/site-packages/ansible/galaxy/collection/__init__.py", line 1328, in _resolve_depenency_map
    return collection_dep_resolver.resolve(
  File "/nix/store/83fcrhal6dkci5gzm6f2kq69zzsrj1l8-python3.8-resolvelib-0.7.1/lib/python3.8/site-packages/resolvelib/resolvers.py", line 472, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/nix/store/83fcrhal6dkci5gzm6f2kq69zzsrj1l8-python3.8-resolvelib-0.7.1/lib/python3.8/site-packages/resolvelib/resolvers.py", line 341, in resolve
    self._add_to_criteria(self.state.criteria, r, parent=None)
  File "/nix/store/83fcrhal6dkci5gzm6f2kq69zzsrj1l8-python3.8-resolvelib-0.7.1/lib/python3.8/site-packages/resolvelib/resolvers.py", line 147, in _add_to_criteria
    matches = self._p.find_matches(
TypeError: find_matches() got an unexpected keyword argument 'identifier'
```

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by dotlambda — CHANGES_REQUESTED** [review]
You have to use `packageOverrides`.

File: pkgs/development/python-modules/ansible/core.nix [hunk 1]
```diff
@@ -22,10 +23,24 @@
 }:
 
 let
+
   ansible-collections = callPackage ./collections.nix {
     version = "4.1.0";
     sha256 = "0rrivq1g0vizah8zmf012lzig2xxfk5x1371k16s3nn4zfkwqqgm";
   };
+
+  # resolvelib 0.x minor versions are breaking.
```
Comment (mweinelt) [inline pkgs/development/python-modules/ansible/core.nix:32]: Wow, that's disappointing. Thanks for reporting this.

**Comment by stephank** [comment]
> You have to use `packageOverrides`.

Sorry, I don't follow, can you expand on this? I only have basic familiarity with Python.

I tried looking for `packageOverrides`, but only see it mentioned in the context of overlays, but I think that's not possible/desired in `python-modules/`? In code, I also cannot find precedent in the `python-modules/` namespace, only in packages outside it that use Python.

**Comment by dotlambda** [comment]
The manual explains how it works indeed. It can be used outside of overlays though, see e.g. https://github.com/NixOS/nixpkgs/blob/9332bcad1813a87d7cd51e94dd9fd4a4aa11093c/pkgs/servers/mautrix-telegram/default.nix.

**Comment by dotlambda** [comment]
> I think that's not possible/desired in `python-modules/`?

Oops, my bad. Overriding Python dependencies of `pythonPackages` is not allowed, sorry.

**Comment by mweinelt** [comment]
We can revert the resolvelib bump, as ansible is its only consumer.

**Comment by stephank** [comment]
> We can revert the resolvelib bump, as ansible is its only consumer.

How will we prevent a bot from breaking this in the future? I'm a little worried that someone/something will create a PR, and it is approved in hasty review (such as during ZHF).

> Oops, my bad. Overriding Python dependencies of `pythonPackages` is not allowed, sorry.

What about adding `resolvelib-0_5` to toplevel pythonPackages instead?

**Comment by mweinelt** [comment]
> How will we prevent a bot from breaking this in the future?

It won't, but I'm the maintainer of the package and I'm now aware of this issue.

> What about adding resolvelib-0_5 to toplevel pythonPackages instead?

As long as Ansible is the only consumer that would result in on unused leaf package. Apart from that we don't support multiple versions of a package within one package set.

**Comment by austinbutler** [comment]
Not sure why this was closed, it seems like a solution was proposed but never implemented as I'm running into the same issue as OP. Created a PR that seems to fix it: https://github.com/NixOS/nixpkgs/pull/133064.

**Pull request closed without merging**